src/v/raft/replicate_batcher.cc

```diff
@@ -54,11 +54,15 @@ replicate_stages replicate_batcher::replicate(
                    }
 
                    enqueued.set_value();
+                   auto item = f.get();
                    return _lock.get_units()
                      .then([this](ss::semaphore_units<> u) {
                          return flush(std::move(u));
                      })
-                     .then([i = f.get()] { return i->_promise.get_future(); });
+                     .handle_exception([item](const std::exception_ptr& e) {
+                         item->_promise.set_exception(e);
+                     })
```

Comment:
seems ok, but this is really hard to analyze and maintain. notice here that if the bottom half of execution in `flush` ever throws or returns an exceptional future and the item's value were set in `flush` because of a non-leader situation, then we'd probably crash-stop here under assertion failure setting promise twice (I assume that is a semantics violation).